File: sdks/node/src/transport.ts
/**
 * How a built batch leaves the worker. The {@link Transport} interface lets
 * tests run the whole pipeline in-process (via {@link FakeTransport}) and lets
 * the daemon / server paths share one worker.
 */

import { endpoint, type Config } from "./config.js";
import type { IngestBatch } from "./wire.js";

/**
 * A transport error. The worker retries once, then drops the batch (the local
 * daemon, in `local_daemon` mode, owns durable retry).
 */
export class TransportError extends Error {
  /** Set when the failure was a non-2xx HTTP status. */
  readonly status?: number;

  constructor(message: string, status?: number) {
    super(message);
    this.name = "TransportError";
    this.status = status;
  }

  /** A non-2xx HTTP status failure. */
  static status(code: number): TransportError {
    return new TransportError(`http status ${code}`, code);
  }

  /** Any other (network / serialization) failure. */
  static other(message: string): TransportError {
    return new TransportError(`transport: ${message}`);
  }
}

/** Ships a built batch to its destination. */
export interface Transport {
  send(batch: IngestBatch): Promise<void>;
}

/** In-memory transport for tests: records every batch it is handed. */
export class FakeTransport implements Transport {
  private readonly recorded: IngestBatch[] = [];

  send(batch: IngestBatch): Promise<void> {
    // Store a structural clone so later mutation of the source can't change a
    // recorded batch (mirrors the Rust `FakeTransport` cloning the batch).
    this.recorded.push(structuredClone(batch));
    return Promise.resolve();
  }

  /** Snapshot of every batch sent so far. */
  batches(): IngestBatch[] {
    return [...this.recorded];
  }
}

/** The real HTTP transport: `POST <endpoint>` with a bearer ingest key. */
export class HttpTransport implements Transport {
  private readonly endpoint: string;
  private readonly bearer: string;

  constructor(endpointUrl: string, bearer: string) {
    this.endpoint = endpointUrl;
    this.bearer = bearer;
  }

  /** Build a transport from a {@link Config}, resolving its mode's endpoint. */
  static fromConfig(cfg: Config): HttpTransport {
    return new HttpTransport(endpoint(cfg.mode), cfg.ingestKey);
  }

  async send(batch: IngestBatch): Promise<void> {
    let resp: Response;
    try {
      resp = await fetch(this.endpoint, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${this.bearer}`,
        },
        body: JSON.stringify(batch),
      });
    } catch (e) {
      // Network-level failure (DNS, connection refused, etc.).
      throw TransportError.other(e instanceof Error ? e.message : String(e));
    }
    if (!resp.ok) {
      throw TransportError.status(resp.status);
    }
  }
}

File: sdks/rust/src/transport.rs
//! How a built batch leaves the worker. The [`Transport`] trait lets tests run
//! the whole pipeline in-process (via [`FakeTransport`]) and lets the daemon /
//! server paths share one worker.

use crate::config::Config;
use crate::wire::IngestBatch;
use async_trait::async_trait;
use std::sync::Mutex;

/// A transport error. The worker retries once, then drops the batch (the local
/// daemon, in `LocalDaemon` mode, owns durable retry).
#[derive(Debug, thiserror::Error)]
pub enum TransportError {
    #[error("http status {0}")]
    Status(u16),
    #[error("transport: {0}")]
    Other(String),
}

/// Ships a built batch to its destination.
#[async_trait]
pub trait Transport: Send + Sync {
    async fn send(&self, batch: &IngestBatch) -> Result<(), TransportError>;
}

/// In-memory transport for tests: records every batch it is handed.
#[derive(Default)]
pub struct FakeTransport {
    batches: Mutex<Vec<IngestBatch>>,
}

impl FakeTransport {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Snapshot of every batch sent so far.
    #[must_use]
    pub fn batches(&self) -> Vec<IngestBatch> {
        self.batches.lock().expect("lock").clone()
    }
}

#[async_trait]
impl Transport for FakeTransport {
    async fn send(&self, batch: &IngestBatch) -> Result<(), TransportError> {
        self.batches.lock().expect("lock").push(batch.clone());
        Ok(())
    }
}

/// The real HTTP transport: `POST <endpoint>` with a bearer ingest key.
pub struct HttpTransport {
    client: reqwest::Client,
    endpoint: String,
    bearer: String,
}

impl HttpTransport {
    #[must_use]
    pub fn from_config(cfg: &Config) -> Self {
        Self {
            client: reqwest::Client::new(),
            endpoint: cfg.mode.endpoint(),
            bearer: cfg.ingest_key.clone(),
        }
    }
}

#[async_trait]
impl Transport for HttpTransport {
    async fn send(&self, batch: &IngestBatch) -> Result<(), TransportError> {
        let resp = self
            .client
            .post(&self.endpoint)
            .bearer_auth(&self.bearer)
            .json(batch)
            .send()
            .await
            .map_err(|e| TransportError::Other(e.to_string()))?;
        let status = resp.status();
        if status.is_success() {
            Ok(())
        } else {
            Err(TransportError::Status(status.as_u16()))
        }
    }
}
